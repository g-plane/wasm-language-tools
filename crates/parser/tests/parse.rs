use codespan_reporting::{
    diagnostic::{Diagnostic, Label},
    files::SimpleFile,
    term::{self, termcolor::Buffer},
};
use insta::{assert_snapshot, glob, Settings};
use std::{fs, path::Path};

#[test]
fn parser_snapshot() {
    glob!("parse/**/*.wat", |path| {
        let input = fs::read_to_string(path).unwrap();
        let (tree, errors) = wat_parser::parse2(&input);
        similar_asserts::assert_eq!(
            tree.to_string(),
            input,
            "syntax tree of '{}' does not match source",
            path.display()
        );

        let file = SimpleFile::new(path.file_name().unwrap().to_str().unwrap(), &input);
        let config = term::Config::default();
        let mut buffer = Buffer::no_color();
        errors
            .into_iter()
            .map(|error| {
                Diagnostic::error()
                    .with_message(error.message.to_string())
                    .with_labels(vec![Label::primary((), error.start..error.end)])
            })
            .for_each(|diagnostic| {
                term::emit(&mut buffer, &config, &file, &diagnostic).unwrap();
            });

        build_settings(path).bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_snapshot!(
                name,
                format!(
                    "{tree:#?}\n{}",
                    String::from_utf8(buffer.into_inner()).unwrap()
                )
            );
        });
    });
}

fn build_settings(path: &Path) -> Settings {
    let mut settings = Settings::clone_current();
    settings.set_snapshot_path(path.parent().unwrap());
    settings.remove_snapshot_suffix();
    settings.set_prepend_module_to_snapshot(false);
    settings.remove_input_file();
    settings.set_omit_expression(true);
    settings.remove_input_file();
    settings.remove_info();
    settings
}
