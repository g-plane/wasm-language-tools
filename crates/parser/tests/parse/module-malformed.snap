---
source: crates/parser/tests/parse.rs
---
ROOT@0..34
  MODULE@0..33
    L_PAREN@0..1 "("
    KEYWORD@1..7 "module"
    WHITESPACE@7..10 "\n  "
    MODULE_FIELD_FUNC@10..16
      L_PAREN@10..11 "("
      KEYWORD@11..15 "func"
      R_PAREN@15..16 ")"
    WHITESPACE@16..19 "\n  "
    ERROR@19..20 "("
    ERROR@20..21 "f"
    ERROR@21..22 ")"
    WHITESPACE@22..25 "\n  "
    MODULE_FIELD_FUNC@25..31
      L_PAREN@25..26 "("
      KEYWORD@26..30 "func"
      R_PAREN@30..31 ")"
    WHITESPACE@31..32 "\n"
    R_PAREN@32..33 ")"
  WHITESPACE@33..34 "\n"

error: unexpected token
  ┌─ module-malformed.wat:3:4
  │
3 │   (f)
  │    ^

error: unexpected token
  ┌─ module-malformed.wat:3:5
  │
3 │   (f)
  │     ^
